c-scape/src/mem/ntbs.rs

```diff
@@ -111,6 +111,134 @@ unsafe extern "C" fn strcmp(mut s1: *const c_char, mut s2: *const c_char) -> c_i
     *s1 as c_uchar as c_int - *s2 as c_uchar as c_int
 }
 
+// enum for strverscmp state
+// internal so no surface
+// see https://codebrowser.dev/glibc/glibc/string/strverscmp.c.html#26
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+enum StrverscmpState {
+    Normal,
+    Integral,
+    Fractional,
+    LeadingZeros,
+}
+
+enum CharType {
+    Zero,
+    Digit,
+    NonNumeric,
+}
+
+impl CharType {
+    fn from_char(c: c_char) -> Self {
+        match c {
+            // ASCII 0
+            48 => Self::Zero,
+            // ASCII 1-9
+            49..=57 => Self::Digit,
+            // non numeric ASCII
+            _ => Self::NonNumeric,
+        }
+    }
+    unsafe fn from_char_raw(c: *const c_char) -> Self {
+        if *c == NUL {
+            return Self::NonNumeric;
+        } else {
+            match *c {
+                // ASCII 0
+                48 => Self::Zero,
+                // ASCII 1-9
+                49..=57 => Self::Digit,
+                // non numeric ASCII
+                _ => Self::NonNumeric,
+            }
+        }
+    }
+}
+
+impl StrverscmpState {
+    fn transition(&mut self, s: CharType) {
+        match s {
+            CharType::Zero => {
+                if self == &Self::Normal {
+                    *self = Self::LeadingZeros;
+                }
+            }
+            // ASCII 1-9
+            CharType::Digit => {
+                if self == &Self::Normal {
+                    *self = Self::Integral;
+                }
+                if self == &Self::LeadingZeros {
+                    *self = Self::Fractional;
+                }
+            }
+            // non numeric ASCII
+            CharType::NonNumeric => *self = Self::Normal,
+        }
+    }
+
+    unsafe fn exit(&mut self, mut s1: *const c_char, mut s2: *const c_char) -> c_int {
+        // safe as loop checks neither pointer is null
```

Comment:
It doesn't appear that `strcasecmp` has defined behavior on null pointers, so this comment shouldn't be needed.